sundials@6.1.1 +rocm build fails: test_nvector.c.o: relocation R_X86_64_32 against symbol `sunctx' can not be used

### Steps to reproduce the issue

`sundials@6.1.1 +rocm` build fails using:
* `spack@develop` (c3b6b3036dede8c14533a1f20ade89a38c926789 from `Mon Mar 14 09:32:09 2022 -0700`)
* Ubuntu 20.04, X86_64
* GCC 9.3.0
* ROCm 4.5.2

Concrete spec: [sundials.spec.json.txt](https://github.com/spack/spack/files/8256612/sundials.spec.json.txt)

Concretization:
```
sundials@6.1.1%gcc@9.3.0+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~caliper~cuda+examples+examples-install~f2003~fcmix+generic-math~hypre~int64~ipo~klu~lapack~monitoring+mpi~openmp~petsc~profiling~pthread~raja+rocm+shared+static~superlu-dist~superlu-mt~sycl~trilinos amdgpu_target=gfx908 build_type=RelWithDebInfo cstd=99 cxxstd=14 precision=double arch=linux-ubuntu20.04-x86_64
    ^cmake@3.22.2%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-x86_64
        ^ncurses@6.2%gcc@9.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
            ^pkgconf@1.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^openssl@1.1.1m%gcc@9.3.0~docs certs=system arch=linux-ubuntu20.04-x86_64
            ^perl@5.34.0%gcc@9.3.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
                ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
                ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
                    ^diffutils@3.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libiconv@1.16%gcc@9.3.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
                ^gdbm@1.19%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^readline@8.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-x86_64
    ^hip@4.5.2%gcc@9.3.0~ipo build_type=Release patches=24ca25a,2a41904 arch=linux-ubuntu20.04-x86_64
        ^comgr@4.5.2%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
            ^llvm-amdgpu@4.5.2%gcc@9.3.0~ipo~link_llvm_dylib~llvm_dylib+openmp+rocm-device-libs build_type=Release patches=d999f3b arch=linux-ubuntu20.04-x86_64
                ^elfutils@0.186%gcc@9.3.0+bzip2~debuginfod~nls+xz arch=linux-ubuntu20.04-x86_64
                    ^m4@1.4.19%gcc@9.3.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
                        ^libsigsegv@2.13%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^xz@5.2.5%gcc@9.3.0+pic libs=shared,static arch=linux-ubuntu20.04-x86_64
                ^hwloc@2.7.0%gcc@9.3.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu20.04-x86_64
                    ^libpciaccess@0.16%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^libtool@2.4.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^util-macros@1.19.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxml2@2.9.12%gcc@9.3.0~python arch=linux-ubuntu20.04-x86_64
                ^perl-data-dumper@2.173%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^python@3.8.12%gcc@9.3.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
                    ^expat@2.4.6%gcc@9.3.0+libbsd arch=linux-ubuntu20.04-x86_64
                        ^libbsd@0.11.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^libmd@1.0.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^gettext@0.21%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-x86_64
                        ^tar@1.34%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libffi@3.4.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^sqlite@3.37.2%gcc@9.3.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-x86_64
                    ^util-linux-uuid@2.37.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^z3@4.8.14%gcc@9.3.0~gmp~ipo~python build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
            ^rocm-cmake@4.5.2%gcc@9.3.0~ipo+ldconfig build_type=Release arch=linux-ubuntu20.04-x86_64
        ^hsa-rocr-dev@4.5.2%gcc@9.3.0+image~ipo+shared build_type=Release patches=71e6851 arch=linux-ubuntu20.04-x86_64
            ^hsakmt-roct@4.5.2%gcc@9.3.0~ipo+shared build_type=Release patches=f926273 arch=linux-ubuntu20.04-x86_64
                ^libdrm@2.4.110%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^docbook-xml@4.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^docbook-xsl@1.79.2%gcc@9.3.0 patches=a92c397 arch=linux-ubuntu20.04-x86_64
                    ^libpthread-stubs@0.4%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^meson@0.61.2%gcc@9.3.0 patches=aa6c50d arch=linux-ubuntu20.04-x86_64
                        ^ninja@1.10.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^py-pip@21.3.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^py-setuptools@59.4.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                            ^py-wheel@0.37.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^numactl@2.0.14%gcc@9.3.0 patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu20.04-x86_64
                    ^autoconf@2.69%gcc@9.3.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
                    ^automake@1.16.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^xxd-standalone@8.2.1201%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^mesa@21.3.1%gcc@9.3.0+glx~llvm+opengl~opengles+osmesa~strip buildtype=debugoptimized default_library=shared patches=ada85be arch=linux-ubuntu20.04-x86_64
            ^bison@3.8.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^flex@2.6.3%gcc@9.3.0+lex~nls arch=linux-ubuntu20.04-x86_64
                ^findutils@4.8.0%gcc@9.3.0 patches=440b954 arch=linux-ubuntu20.04-x86_64
            ^glproto@1.4.17%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libunwind@1.5.0%gcc@9.3.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz~zlib components=none libs=shared,static arch=linux-ubuntu20.04-x86_64
            ^libx11@1.7.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^inputproto@2.3.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^kbproto@1.0.7%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libxcb@1.14%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxau@1.0.8%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^xproto@7.0.31%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxdmcp@1.1.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^xcb-proto@1.14.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^xextproto@7.3.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^xtrans@1.3.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libxext@1.3.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^libxt@1.1.5%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libice@1.0.9%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libsm@1.2.3%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^py-mako@1.1.6%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^py-markupsafe@2.0.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
            ^xrandr@1.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                ^libxrandr@1.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^libxrender@0.9.10%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                        ^renderproto@0.11.1%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
                    ^randrproto@1.5.0%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
        ^rocminfo@4.5.2%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-x86_64
        ^roctracer-dev-api@4.5.2%gcc@9.3.0 arch=linux-ubuntu20.04-x86_64
    ^mpich@3.4.2%gcc@9.3.0~argobots+fortran+hwloc+hydra+libxml2+pci+romio~slurm~two_level_namespace~verbs~wrapperrpath device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
        ^libfabric@1.14.0%gcc@9.3.0~debug~kdreg fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
```

Error is quickly reproduced using Docker container image and cached dependency binaries:
* Docker image = `ecpe4s/ubuntu20.04-runner-x86_64:2022-02-01`
* Cached dependencies @ `https://cache.e4s.io`

```
root@d2bc3d9aa585:/# git clone https://github.com/spack/spack
root@d2bc3d9aa585:/# (cd. spack && git checkout c3b6b3036)

root@d2bc3d9aa585:/# . spack/share/spack/setup-env.sh
root@d2bc3d9aa585:/# spack mirror add E4S https://cache.e4s.io
root@d2bc3d9aa585:/# spack buildcache keys -it

root@d2bc3d9aa585:/# spack install --cache-only --include-build-deps --only dependencies -f ./sundials.spec.json

root@d2bc3d9aa585:/# spack install --no-cache --only package -f ./sundials.spec.json
...
==> Installing sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633
==> No binary for sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633 found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/cf/cfaf637b792c330396a25ef787eb59d58726c35918ebbc08e33466e45d50470c.tar.gz
==> No patches needed for sundials
==> sundials: Executing phase: 'cmake'
==> sundials: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

13 errors found in build log:
     3308    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.4.2-b2hsqqqnsiodzp3kw63u425osorilfty/bin/mpicc -DLLVM_VERSION_SUFFIX=git -D_POSIX_C_SOURCE=199309L -O2 -g -D
             NDEBUG CMakeFiles/kinFoodWeb_kry_p.dir/kinFoodWeb_kry_p.c.o -o kinFoodWeb_kry_p  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha
             6j633/spack-build-7embres/src/kinsol:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/parallel ../../..
             /src/kinsol/libsundials_kinsol.so.6.1.1 ../../../src/nvector/parallel/libsundials_nvecparallel.so.6.1.1 -lm
     3309    [ 89%] Linking C executable ark_analytic_nonlin
     3310    cd /tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/examples/arkode/C_serial && /spack/opt/spack/linux-ubuntu20.04-
             x86_64/gcc-9.3.0/cmake-3.22.2-4oedsu6d3q7kirp2izapy5ygl3cfqsic/bin/cmake -E cmake_link_script CMakeFiles/ark_analytic_nonlin.dir/link.txt --verbose=1
     3311    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/llvm-amdgpu-4.5.2-a3yp5hsiusj5w5qp4ovtlybyojkj5zj4/bin/clang -DLLVM_VERSION_SUFFIX=git -D_POSIX_C_SOURCE=199309L -O2
              -g -DNDEBUG CMakeFiles/ark_analytic_nonlin.dir/ark_analytic_nonlin.c.o -o ark_analytic_nonlin  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5
             w3w3ozbz7jufdha6j633/spack-build-7embres/src/arkode:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/se
             rial ../../../src/arkode/libsundials_arkode.so.5.1.1 ../../../src/nvector/serial/libsundials_nvecserial.so.6.1.1 -lm
     3312    /usr/bin/ld: ../CMakeFiles/test_nvector_obj.dir/test_nvector.c.o: relocation R_X86_64_32 against symbol `sunctx' can not be used when making a PIE object; recompile wit
             h -fPIE
     3313    /usr/bin/ld: ../CMakeFiles/test_nvectormpi_obj.dir/test_mpinvector.c.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recomp
             ile with -fPIE
  >> 3314    collect2: error: ld returned 1 exit status
     3315    [ 89%] Linking C executable ark_brusselator_fp
     3316    [ 89%] Linking C executable ark_analytic
     3317    [ 90%] Linking C executable ark_analytic_mels
     3318    cd /tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/examples/arkode/C_serial && /spack/opt/spack/linux-ubuntu20.04-
             x86_64/gcc-9.3.0/cmake-3.22.2-4oedsu6d3q7kirp2izapy5ygl3cfqsic/bin/cmake -E cmake_link_script CMakeFiles/ark_brusselator_fp.dir/link.txt --verbose=1
     3319    cd /tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/examples/arkode/C_serial && /spack/opt/spack/linux-ubuntu20.04-
             x86_64/gcc-9.3.0/cmake-3.22.2-4oedsu6d3q7kirp2izapy5ygl3cfqsic/bin/cmake -E cmake_link_script CMakeFiles/ark_analytic.dir/link.txt --verbose=1
     3320    cd /tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/examples/arkode/C_serial && /spack/opt/spack/linux-ubuntu20.04-
             x86_64/gcc-9.3.0/cmake-3.22.2-4oedsu6d3q7kirp2izapy5ygl3cfqsic/bin/cmake -E cmake_link_script CMakeFiles/ark_analytic_mels.dir/link.txt --verbose=1
  >> 3321    make[2]: *** [examples/nvector/parallel/CMakeFiles/test_nvector_mpi.dir/build.make:107: examples/nvector/parallel/test_nvector_mpi] Error 1
     3322    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
  >> 3323    make[1]: *** [CMakeFiles/Makefile2:10000: examples/nvector/parallel/CMakeFiles/test_nvector_mpi.dir/all] Error 2
     3324    make[1]: *** Waiting for unfinished jobs....
     3325    [ 91%] Linking C executable ark_brusselator
     3326    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/llvm-amdgpu-4.5.2-a3yp5hsiusj5w5qp4ovtlybyojkj5zj4/bin/clang -DLLVM_VERSION_SUFFIX=git -D_POSIX_C_SOURCE=199309L -O2
              -g -DNDEBUG CMakeFiles/ark_analytic.dir/ark_analytic.c.o -o ark_analytic  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/
             spack-build-7embres/src/arkode:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/serial ../../../src/ark
             ode/libsundials_arkode.so.5.1.1 ../../../src/nvector/serial/libsundials_nvecserial.so.6.1.1 -lm
     3327    cd /tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/examples/arkode/C_serial && /spack/opt/spack/linux-ubuntu20.04-
             x86_64/gcc-9.3.0/cmake-3.22.2-4oedsu6d3q7kirp2izapy5ygl3cfqsic/bin/cmake -E cmake_link_script CMakeFiles/ark_brusselator.dir/link.txt --verbose=1
     3328    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/llvm-amdgpu-4.5.2-a3yp5hsiusj5w5qp4ovtlybyojkj5zj4/bin/clang -DLLVM_VERSION_SUFFIX=git -D_POSIX_C_SOURCE=199309L -O2
              -g -DNDEBUG CMakeFiles/ark_brusselator_fp.dir/ark_brusselator_fp.c.o -o ark_brusselator_fp  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w
             3ozbz7jufdha6j633/spack-build-7embres/src/arkode:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/seria
             l ../../../src/arkode/libsundials_arkode.so.5.1.1 ../../../src/nvector/serial/libsundials_nvecserial.so.6.1.1 -lm
     3329    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/llvm-amdgpu-4.5.2-a3yp5hsiusj5w5qp4ovtlybyojkj5zj4/bin/clang -DLLVM_VERSION_SUFFIX=git -D_POSIX_C_SOURCE=199309L -O2
              -g -DNDEBUG CMakeFiles/ark_analytic_mels.dir/ark_analytic_mels.c.o -o ark_analytic_mels  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3oz
             bz7jufdha6j633/spack-build-7embres/src/arkode:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/serial .
             ./../../src/arkode/libsundials_arkode.so.5.1.1 ../../../src/nvector/serial/libsundials_nvecserial.so.6.1.1 -lm
     3330    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/llvm-amdgpu-4.5.2-a3yp5hsiusj5w5qp4ovtlybyojkj5zj4/bin/clang -DLLVM_VERSION_SUFFIX=git -D_POSIX_C_SOURCE=199309L -O2
              -g -DNDEBUG CMakeFiles/ark_brusselator.dir/ark_brusselator.c.o -o ark_brusselator  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7juf
             dha6j633/spack-build-7embres/src/arkode:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/serial ../../.
             ./src/arkode/libsundials_arkode.so.5.1.1 ../../../src/nvector/serial/libsundials_nvecserial.so.6.1.1 -lm
     3331    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
     3332    /usr/bin/ld: ../CMakeFiles/test_nvector_obj.dir/test_nvector.c.o: relocation R_X86_64_32 against symbol `sunctx' can not be used when making a PIE object; recompile wit
             h -fPIE
     3333    /usr/bin/ld: ../CMakeFiles/test_nvectormpi_obj.dir/test_mpinvector.c.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recomp
             ile with -fPIE
  >> 3334    collect2: error: ld returned 1 exit status
  >> 3335    make[2]: *** [examples/nvector/mpiplusx/CMakeFiles/test_nvector_mpiplusx.dir/build.make:108: examples/nvector/mpiplusx/test_nvector_mpiplusx] Error 1
     3336    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
  >> 3337    make[1]: *** [CMakeFiles/Makefile2:10090: examples/nvector/mpiplusx/CMakeFiles/test_nvector_mpiplusx.dir/all] Error 2
     3338    [ 91%] Built target ark_analytic_nonlin
     3339    /usr/bin/ld: ../CMakeFiles/test_nvector_obj.dir/test_nvector.c.o: relocation R_X86_64_32 against symbol `sunctx' can not be used when making a PIE object; recompile wit
             h -fPIE
     3340    /usr/bin/ld: ../CMakeFiles/test_nvectormpi_obj.dir/test_mpinvector.c.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recomp
             ile with -fPIE
  >> 3341    collect2: error: ld returned 1 exit status
  >> 3342    make[2]: *** [examples/nvector/mpimanyvector/CMakeFiles/test_nvector_mpimanyvector_parallel1.dir/build.make:108: examples/nvector/mpimanyvector/test_nvector_mpimanyvect
             or_parallel1] Error 1
     3343    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
  >> 3344    make[1]: *** [CMakeFiles/Makefile2:10030: examples/nvector/mpimanyvector/CMakeFiles/test_nvector_mpimanyvector_parallel1.dir/all] Error 2
     3345    /usr/bin/ld: ../CMakeFiles/test_nvector_obj.dir/test_nvector.c.o: relocation R_X86_64_32 against symbol `sunctx' can not be used when making a PIE object; recompile wit
             h -fPIE
     3346    /usr/bin/ld: ../CMakeFiles/test_nvectormpi_obj.dir/test_mpinvector.c.o: relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a PIE object; recomp
             ile with -fPIE
  >> 3347    collect2: error: ld returned 1 exit status
     3348    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
  >> 3349    make[2]: *** [examples/nvector/mpimanyvector/CMakeFiles/test_nvector_mpimanyvector_parallel2.dir/build.make:108: examples/nvector/mpimanyvector/test_nvector_mpimanyvect
             or_parallel2] Error 1
     3350    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
  >> 3351    make[1]: *** [CMakeFiles/Makefile2:10060: examples/nvector/mpimanyvector/CMakeFiles/test_nvector_mpimanyvector_parallel2.dir/all] Error 2
     3352    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
     3353    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
     3354    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
     3355    [ 91%] Built target ark_analytic
     3356    [ 91%] Built target kinFoodWeb_kry_bbd_p
     3357    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'

     ...

     3377    [ 91%] Linking CXX executable kin_heat2D_nonlin_p
     3378    cd /tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/examples/kinsol/CXX_parallel && /spack/opt/spack/linux-ubuntu20
             .04-x86_64/gcc-9.3.0/cmake-3.22.2-4oedsu6d3q7kirp2izapy5ygl3cfqsic/bin/cmake -E cmake_link_script CMakeFiles/kin_heat2D_nonlin_p.dir/link.txt --verbose=1
     3379    /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/hip-4.5.2-tawzuyewylqa3yvzzn45eet237thos3x/bin/hipcc -DLLVM_VERSION_SUFFIX=git -O2 -g -DNDEBUG -Wl,-rpath -Wl,/spack
             /opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.4.2-b2hsqqqnsiodzp3kw63u425osorilfty/lib -L/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/hwloc-2.7.0-hswya4s
             r4io27hwpjsb7oyhgqb6vzy34/lib CMakeFiles/kin_heat2D_nonlin_p.dir/kin_heat2D_nonlin_p.cpp.o -o kin_heat2D_nonlin_p  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-sundials
             -6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/kinsol:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7em
             bres/src/nvector/parallel:/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/mpiplusx:/tmp/root/spack-sta
             ge/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres/src/nvector/serial:/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.4.2-b2
             hsqqqnsiodzp3kw63u425osorilfty/lib ../../../src/kinsol/libsundials_kinsol.so.6.1.1 ../../../src/nvector/parallel/libsundials_nvecparallel.so.6.1.1 ../../../src/nvector/
             mpiplusx/libsundials_nvecmpiplusx.so.6.1.1 ../../../src/nvector/serial/libsundials_nvecserial.so.6.1.1 -lm /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.4
             .2-b2hsqqqnsiodzp3kw63u425osorilfty/lib/libmpicxx.so /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.3.0/mpich-3.4.2-b2hsqqqnsiodzp3kw63u425osorilfty/lib/libmpi.so
     3380    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
     3381    [ 91%] Built target kin_heat2D_nonlin_p
     3382    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-sundials-6.1.1-7embresvuen5w3w3ozbz7jufdha6j633/spack-build-7embres'
  >> 3383    make: *** [Makefile:149: all] Error 2
```

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>

### Information on your system

* **Spack:** 0.17.1-1549-c3b6b3036d
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/8256610/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/8256620/spack-build-out.txt)

@wspear
@balos1 
@cswoodward 
@gardner48


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml
@@ -199,6 +199,7 @@ spack:
   - slate +rocm amdgpu_target=gfx90a
   - slepc +rocm amdgpu_target=gfx90a ^petsc +rocm amdgpu_target=gfx90a
   - strumpack ~slate +rocm amdgpu_target=gfx90a
+  - sundials +rocm amdgpu_target=gfx90a
   - superlu-dist +rocm amdgpu_target=gfx90a
   - tasmanian ~openmp +rocm amdgpu_target=gfx90a
   - tau +mpi +rocm

--- a/var/spack/repos/builtin/packages/sundials/nvector-pic.patch
+++ b/var/spack/repos/builtin/packages/sundials/nvector-pic.patch
@@ -1,0 +1,24 @@
+diff -ruN spack-src/examples/nvector/CMakeLists.txt spack-src-patched/examples/nvector/CMakeLists.txt
+--- spack-src/examples/nvector/CMakeLists.txt	2022-04-22 16:55:14.000000000 +0000
++++ spack-src-patched/examples/nvector/CMakeLists.txt	2022-08-03 16:34:47.789320825 +0000
+@@ -25,6 +25,10 @@
+ 
+ # Build the nvector test utilities
+ add_library(test_nvector_obj OBJECT test_nvector.c)
++if(BUILD_SHARED_LIBS)
++  # need PIC when shared libs are used since the example executables will link to the shared lib
++  set_property(TARGET test_nvector_obj PROPERTY POSITION_INDEPENDENT_CODE TRUE)
++endif()
+ target_link_libraries(test_nvector_obj PRIVATE sundials_nvecserial)
+ 
+ if(ENABLE_MPI AND MPI_C_FOUND)
+@@ -33,6 +37,9 @@
+   add_subdirectory(mpiplusx)
+   # Build the mpi nvector test utilities
+   add_library(test_nvectormpi_obj OBJECT test_mpinvector.c)
++  if(BUILD_SHARED_LIBS)
++    set_property(TARGET test_nvectormpi_obj PROPERTY POSITION_INDEPENDENT_CODE TRUE)
++  endif()
+   target_link_libraries(test_nvectormpi_obj PRIVATE MPI::MPI_C sundials_nvecparallel)
+ endif()
+ add_subdirectory(manyvector)

--- a/var/spack/repos/builtin/packages/sundials/package.py
+++ b/var/spack/repos/builtin/packages/sundials/package.py
@@ -237,6 +237,8 @@ class Sundials(CMakePackage, CudaPackage, ROCmPackage):
     # ==========================================================================
     # Patches
     # ==========================================================================
+    # https://github.com/spack/spack/issues/29526
+    patch("nvector-pic.patch", when="@6.1.0:6.2.0 +rocm")
 
     # remove OpenMP header file and function from hypre vector test code
     patch("test_nvector_parhyp.patch", when="@2.7.0:3.0.0")

